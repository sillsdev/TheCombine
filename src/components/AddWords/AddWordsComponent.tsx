import React from "react";
import {
  Paper,
  Typography,
  Container,
  TextField,
  Grid,
  Divider,
  Button,
  IconButton,
  Tooltip,
  Chip,
  Dialog
} from "@material-ui/core";
import theme from "../../types/theme";

import { Translate, TranslateFunction } from "react-localize-redux";
import { Word, SemanticDomain, State, Gloss } from "../../types/word";
import { Delete, Edit } from "@material-ui/icons";
import * as Backend from "../../backend";
import DuplicateFinder from "../../goals/MergeDupGoal/DuplicateFinder/DuplicateFinder";
import DomainTree from "../TreeView/SemanticDomain";
import TreeViewComponent from "../TreeView";

interface AddWordsProps {
  domain: DomainTree;
  translate: TranslateFunction;
}

interface AddWordsState {
  rows: Row[];
  newVern: string;
  newGloss: string;
  hoverRow?: number;
  newVernInFrontier: Boolean; // does the new word already exist in the frontier?
  showDuplicate?: number;
  gettingSemanticDomain: boolean;
}

/** A row in the view */
interface Row {
  id: string;
  /** The ID of the duplicate word in the frontier */
  dupId: string;
  vernacular: string;
  glosses: string;
  /** The index of the sense of the word that we're showing in the view */
  senseIndex: number;
  /** The vernacular of the duplicate word in the frontier */
  dupVernacular?: string;
  /** The `def` field of the glosses of the duplicate word in the frontier */
  dupGlosses?: string[];
}

export default class AddWords extends React.Component<
  AddWordsProps,
  AddWordsState
> {
  constructor(props: AddWordsProps) {
    super(props);
    this.state = {
      newVern: "",
      newGloss: "",
      rows: [],
      newVernInFrontier: false,
      gettingSemanticDomain: true
    };
    this.vernInput = React.createRef<HTMLDivElement>();
    this.glossInput = React.createRef<HTMLDivElement>();
  }

  allWords: Word[] = [];
  semanticDomain: SemanticDomain = { name: "Sky", id: "1.2" };

  async componentDidMount() {
    this.allWords = await Backend.getFrontierWords();
  }

  /** If the venacular is in the frontier, returns that words id */
  vernInFrontier(vernacular: string): string {
    let Finder = new DuplicateFinder();

    //[vernacular form, levenshtein distance]
    // the number defined here sets the upper bound on acceptable scores
    let foundDuplicate: [string, number] = ["", 2];

    for (let word of this.allWords) {
      let accessible = false;
      for (let sense of word.senses) {
        if (sense.accessibility === 0) {
          accessible = true;
          break;
        }
      }
      if (accessible) {
        let levenD: number = Finder.getLevenshteinDistance(
          vernacular,
          word.vernacular
        );
        if (levenD < foundDuplicate[1]) {
          foundDuplicate = [word.id, levenD];
        }
      }
    }

    return foundDuplicate[0];
  }

  vernInput: React.RefObject<HTMLDivElement>;
  glossInput: React.RefObject<HTMLDivElement>;

  submit(e?: React.FormEvent<HTMLFormElement>, callback?: Function) {
    if (e) e.preventDefault();

    const vernacular = this.state.newVern;
    const glosses = this.state.newGloss;

    if (vernacular === "") return;

    let rows = [...this.state.rows];
    Backend.createWord(
      this.rowToNewWord({
        vernacular,
        glosses,
        id: "",
        dupId: "",
        senseIndex: 0
      })
    )
      .catch(err => console.log(err))
      .then(res => {
        let word = res as Word;
        let dupId = this.vernInFrontier(word.vernacular);
        rows.push({ ...this.wordToRow(word, 0), dupId });
        this.setState({
          rows,
          newVern: "",
          newGloss: "",
          newVernInFrontier: false
        });
        this.focusVernInput();
        if (callback) callback(res);
      });
  }

  /** Return a new word based on a row */
  rowToNewWord(row: Row): Word {
    let word: Word = {
      id: row.id,
      vernacular: "",
      senses: [
        {
          glosses: [],
          semanticDomains: [
            { name: this.props.domain.name, id: this.props.domain.id }
          ]
        }
      ],
      audio: [],
      created: "",
      modified: "",
      history: [],
      partOfSpeech: "",
      editedBy: [],
      otherField: "",
      plural: ""
    };
    word.vernacular = row.vernacular;

    word.senses[0].glosses = [];
    let defs = row.glosses.split(",");
    for (let def of defs) {
      let gloss = {
        language: "en",
        def: def.trim()
      };
      word.senses[0].glosses.push(gloss);
    }

    return word;
  }

  /** Update a row in the view only */
  updateRow(row: Row, index: number, callback?: Function) {
    let rows = [...this.state.rows];
    rows.splice(index, 1, { ...rows[index], ...row });
    if (callback) this.setState({ rows }, () => callback());
    else this.setState({ rows });
  }

  /** Update the word in the backend */
  updateWord(index: number, callback?: Function) {
    let row = this.state.rows[index];
    this.rowToExistingWord(row)
      .catch(err => console.log(err))
      .then(res =>
        Backend.updateWord(res as Word)
          .catch(err => console.log(err))
          .then(res => {
            this.updateRow(this.wordToRow(res as Word, row.senseIndex), index);
            if (callback) callback();
          })
      );
  }

  /** Add the fields in a row to the word it corresponds to in the database */
  async rowToExistingWord(row: Row): Promise<Word> {
    let word = await Backend.getWord(row.id);

    let glosses: Gloss[] = [];
    let gloss: Gloss;
    let defs = row.glosses.split(",");
    for (let def of defs) {
      gloss = {
        language: "en",
        def: def.trim()
      };
      glosses.push(gloss);
    }
    word.senses[row.senseIndex] = {
      glosses,
      semanticDomains: [this.semanticDomain]
    };

    word.vernacular = row.vernacular;
    return word;
  }

  // Used by new word input
  /** Update the state to match the value in a textbox */
  updateField<K extends keyof AddWordsState>(
    e: React.ChangeEvent<
      HTMLTextAreaElement | HTMLInputElement | HTMLSelectElement
    >,
    field: K,
    callback?: () => void
  ) {
    const value = e.target.value;

    this.setState(
      {
        [field]: value
      } as Pick<AddWordsState, K>,
      callback
    );
  }

  /** Move the focus to the vernacular textbox */
  focusVernInput() {
    if (this.vernInput.current) this.vernInput.current.focus();
  }

  /** Move the focus to the gloss textbox */
  focusGlossInput() {
    if (this.glossInput.current) this.glossInput.current.focus();
  }

  /** Remove a word from the backend */
  removeWord(id: string, callback?: Function) {
    Backend.deleteWordById(id)
      .catch(err => console.log(err))
      .then(res => {
        if (callback) callback(res);
      });
  }

  /** Delete a row from the view only */
  removeRow(index: number) {
    let rows = [...this.state.rows];
    rows.splice(index, 1);
    this.setState({ rows });
  }

  wordToRow(word: Word, senseIndex: number): Row {
    let row: Row = {
      vernacular: word.vernacular,
      id: word.id,
      glosses: "",
      dupId: "",
      senseIndex
    };
    let glosses: string[] = [];
    word.senses[senseIndex].glosses.forEach(gloss => {
      glosses.push(gloss.def);
    });
    row.glosses = glosses.join(", ");
    return row;
  }

  getWord(id: string): Word {
    let word = this.allWords.find(word => word.id === id);
    if (!word) throw new Error("No word exists with this id");
    return word;
  }

  showDuplicateForRow(rowIndex: number) {
    let row = this.state.rows[rowIndex];
    let dupWord = this.getWord(row.dupId);
    row.dupVernacular = dupWord.vernacular;
    row.dupGlosses = [];
    for (let sense of dupWord.senses) {
      let glosses = [];
      for (let gloss of sense.glosses) {
        glosses.push(gloss.def);
      }
      row.dupGlosses.push(glosses.join(", "));
    }
    this.updateRow(row, rowIndex);
    this.setState({ showDuplicate: rowIndex });
  }

  /** Switch a row to edit a sense of a word in the database */
  switchToExistingWord(rowIndex: number, senseIndex: number) {
    let row = this.state.rows[rowIndex];
    if (row.dupId === "") {
      throw new Error("This row does not have a duplicate");
    } else {
      Backend.deleteWordById(row.id)
        .catch(err => console.log(err))
        .then(() => {
          if (row.dupVernacular && row.dupGlosses) {
            let newRow: Row = {
              vernacular: row.dupVernacular,
              glosses: row.dupGlosses[senseIndex]
                ? row.dupGlosses[senseIndex]
                : row.glosses,
              id: row.dupId,
              dupId: "",
              senseIndex
            };
            this.updateRow(newRow, rowIndex, () => this.updateWord(rowIndex));
          }
        });
    }
  }

  render() {
    return (
      <Container>
        <Paper
          style={{
            padding: theme.spacing(2),
            maxWidth: 800,
            marginLeft: "auto",
            marginRight: "auto"
          }}
        >
          <Typography
            variant="h4"
            align="center"
            style={{ marginBottom: theme.spacing(2) }}
          >
            <Translate id="addWords.domain" />
            {": "}
            {this.props.domain.name + " (" + this.props.domain.id + ")"}
            <IconButton
              onClick={() => {
                this.setState({ gettingSemanticDomain: true });
              }}
            >
              <Edit />
            </IconButton>
          </Typography>
          <Divider />
          <form onSubmit={e => this.submit(e)}>
            <input type="submit" style={{ display: "none" }} />

            {/* Table title */}
            <Grid container spacing={3}>
              <Grid item xs={5}>
                <Typography
                  variant="h5"
                  align="center"
                  style={{ marginTop: theme.spacing(2) }}
                >
                  <Translate id="addWords.vernacular" />
                </Typography>
              </Grid>
              <Grid item xs={5}>
                <Typography
                  variant="h5"
                  align="center"
                  style={{ marginTop: theme.spacing(2) }}
                >
                  <Translate id="addWords.glosses" />
                </Typography>
              </Grid>

              {/* Rows of words */}
              {this.state.rows.map((row, rowIndex) => {
                return (
                  <React.Fragment>
                    <Grid
                      item
                      xs={12}
                      key={rowIndex}
                      onMouseEnter={() => this.setState({ hoverRow: rowIndex })}
                      onMouseLeave={() =>
                        this.setState({ hoverRow: undefined })
                      }
                    >
                      <Grid container>
                        <Grid
                          item
                          xs={5}
                          style={{
                            paddingLeft: theme.spacing(2),
                            paddingRight: theme.spacing(2),
                            position: "relative"
                          }}
                        >
                          <TextField
                            fullWidth
                            value={row.vernacular}
                            onChange={e => {
                              let dupId = this.vernInFrontier(e.target.value);
                              if (dupId === row.id) dupId = ""; // the "duplicate" is the word we're already editing
                              this.updateRow(
                                {
                                  ...row,
                                  vernacular: e.target.value,
                                  dupId
                                },
                                rowIndex
                              );
                            }}
                            onBlur={() => {
                              this.updateWord(rowIndex);
                            }}
                            onKeyDown={e => {
                              if (e.key === "Enter") {
                                this.focusVernInput();
                              }
                            }}
                          />
                          {row.dupId !== "" && (
                            <Tooltip
                              title={
                                this.props.translate(
                                  "addWords.wordInDatabase"
                                ) as string
                              }
                              placement="top"
                            >
                              <div
                                style={{
                                  height: "5px",
                                  width: "5px",
                                  border: "2px solid red",
                                  borderRadius: "50%",
                                  position: "absolute",
                                  top: 8,
                                  right: 48,
                                  cursor: "pointer"
                                }}
                                onClick={() =>
                                  this.showDuplicateForRow(rowIndex)
                                }
                              />
                            </Tooltip>
                          )}
                        </Grid>
                        <Grid
                          item
                          xs={5}
                          style={{
                            paddingLeft: theme.spacing(2),
                            paddingRight: theme.spacing(2)
                          }}
                        >
                          <TextField
                            fullWidth
                            value={row.glosses}
                            onChange={e => {
                              this.updateRow(
                                { ...row, glosses: e.target.value },
                                rowIndex
                              );
                            }}
                            onBlur={() => {
                              this.updateWord(rowIndex);
                            }}
                            onKeyDown={e => {
                              if (e.key === "Enter") {
                                this.focusVernInput();
                              }
                            }}
                          />
                        </Grid>
                        <Grid item xs={2}>
                          {this.state.hoverRow === rowIndex && (
                            <React.Fragment>
                              <Tooltip
                                title={
                                  this.props.translate(
                                    "addWords.deleteRow"
                                  ) as string
                                }
                                placement="top"
                              >
                                <IconButton
                                  size="small"
                                  onClick={() =>
                                    this.removeWord(row.id, () =>
                                      this.removeRow(rowIndex)
                                    )
                                  }
                                >
                                  <Delete />
                                </IconButton>
                              </Tooltip>
                            </React.Fragment>
                          )}
                        </Grid>
                      </Grid>
                    </Grid>
                    {/* This is where it shows the duplicate if the red dot is clicked */}
                    {this.state.showDuplicate === rowIndex && row.dupId && (
                      <Grid
                        item
                        xs={12}
                        key={"d" + rowIndex}
                        onMouseEnter={() =>
                          this.setState({ hoverRow: rowIndex })
                        }
                        onMouseLeave={() =>
                          this.setState({ hoverRow: undefined })
                        }
                        style={{ background: "whitesmoke" }}
                      >
                        <Grid container>
                          <Grid
                            item
                            xs={5}
                            style={{
                              paddingLeft: theme.spacing(2),
                              paddingRight: theme.spacing(2)
                            }}
                          >
                            <Typography variant="body1">
                              {"Duplicate in database: " + row.dupVernacular}
                            </Typography>
                          </Grid>
                          <Grid
                            item
                            xs={5}
                            style={{
                              paddingLeft: theme.spacing(2),
                              paddingRight: theme.spacing(2)
                            }}
                          >
                            <Typography variant="body1">
                              {"Glosses: "}
                            </Typography>
                            {row.dupGlosses &&
                              row.dupGlosses.map((gloss, senseIndex) => (
                                <Chip
                                  label={gloss}
                                  onClick={() =>
                                    this.switchToExistingWord(
                                      rowIndex,
                                      senseIndex
                                    )
                                  }
                                  style={{
                                    margin: 4
                                  }}
                                />
                              ))}
                            <Chip
                              variant="outlined"
                              label={"Add New Sense +"}
                              onClick={() =>
                                this.switchToExistingWord(
                                  rowIndex,
                                  row.dupGlosses ? row.dupGlosses.length : 0
                                )
                              }
                              style={{
                                margin: 4
                              }}
                            />
                          </Grid>
                        </Grid>
                      </Grid>
                    )}
                  </React.Fragment>
                );
              })}

              {/* New word entry */}
              <React.Fragment>
                <Grid item xs={12}>
                  <Grid container>
                    <Grid
                      item
                      xs={5}
                      style={{
                        paddingLeft: theme.spacing(2),
                        paddingRight: theme.spacing(2),
                        position: "relative"
                      }}
                    >
                      <TextField
                        autoFocus
                        label={<Translate id="addWords.vernacular" />}
                        fullWidth
                        variant="outlined"
                        value={this.state.newVern}
                        onChange={e => {
                          this.updateField(e, "newVern", () =>
                            this.setState({
                              newVernInFrontier:
                                this.vernInFrontier(this.state.newVern) !== ""
                            })
                          );
                        }}
                        inputRef={this.vernInput}
                        // Move the focus to the next box when the right arrow key is pressed
                        onKeyDown={e => {
                          if (
                            e.key === "ArrowRight" &&
                            (e.target as HTMLInputElement).selectionStart ===
                              this.state.newVern.length
                          )
                            this.focusGlossInput();
                        }}
                      />
                      {this.state.newVernInFrontier && (
                        <Tooltip
                          title={
                            this.props.translate(
                              "addWords.wordInDatabase"
                            ) as string
                          }
                          placement="top"
                        >
                          <div
                            style={{
                              height: "5px",
                              width: "5px",
                              border: "2px solid red",
                              borderRadius: "50%",
                              position: "absolute",
                              top: 24,
                              right: 48,
                              cursor: "pointer"
                            }}
                          />
                        </Tooltip>
                      )}
                    </Grid>
                    <Grid
                      item
                      xs={5}
                      style={{
                        paddingLeft: theme.spacing(2),
                        paddingRight: theme.spacing(2)
                      }}
                    >
                      <TextField
                        label={<Translate id="addWords.glosses" />}
                        fullWidth
                        variant="outlined"
                        value={this.state.newGloss}
                        onChange={e => {
                          this.updateField(e, "newGloss");
                        }}
                        inputRef={this.glossInput}
                        // Move the focus to the previous box when the left arrow key is pressed
                        onKeyDown={e => {
                          if (
                            e.key === "ArrowLeft" &&
                            (e.target as HTMLInputElement).selectionStart === 0
                          )
                            this.focusVernInput();
                        }}
                      />
                    </Grid>
                  </Grid>
                </Grid>
              </React.Fragment>
            </Grid>
          </form>
          <Grid container justify="flex-end" spacing={2}>
            <Grid item>
              <Button
                type="submit"
                variant="contained"
                color="primary"
                style={{ marginTop: theme.spacing(2) }}
              >
                <Translate id="addWords.next" />
              </Button>
            </Grid>
          </Grid>
        </Paper>

        {/** Tree modal */}
        <Dialog fullScreen open={this.state.gettingSemanticDomain}>
          <TreeViewComponent
            returnControlToCaller={() =>
              this.setState({
                gettingSemanticDomain: false
              })
            }
          />
        </Dialog>
      </Container>
    );
  }
}
