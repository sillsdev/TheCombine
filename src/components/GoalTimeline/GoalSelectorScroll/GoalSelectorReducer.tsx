import { StoreAction, StoreActions } from "../../../rootActions";
import { GoalSelectorState } from "../../../types/goals";
import {
  GoalScrollAction,
  MOUSE_ACTION,
  SELECT_ACTION,
} from "./GoalSelectorAction";

export const defaultState: GoalSelectorState = {
  selectedIndex: 0,
  allPossibleGoals: [],
  mouseX: 0,
  lastIndex: 0,
};

export const goalSelectReducer = (
  state: GoalSelectorState | undefined,
  action: StoreAction | GoalScrollAction
): GoalSelectorState => {
  if (!state) return defaultState;
  switch (action.type) {
    case SELECT_ACTION:
      return {
        ...state,
        selectedIndex: action.payload,
      };
    case MOUSE_ACTION:
      return {
        ...state,
        mouseX: action.payload,
      };
    case StoreActions.RESET:
      return defaultState;
    default:
      return state;
  }
};
