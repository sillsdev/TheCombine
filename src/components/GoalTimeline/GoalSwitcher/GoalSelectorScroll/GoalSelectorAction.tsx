export const SELECT_ACTION = "PRESS_BUTTON";
export type SELECT_ACTION = typeof SELECT_ACTION;

export const MOUSE_ACTION = "MOUSE_ACTION";
export type MOUSE_ACTION = typeof MOUSE_ACTION;

//action types
export interface ScrollSelectorAct {
  type: SELECT_ACTION;
  payload: number;
}
export interface MouseMoveAct {
  type: MOUSE_ACTION;
  payload: number;
}

export function scrollSelectorIndexAction(
  selectedIndex: number
): ScrollSelectorAct {
  return {
    type: SELECT_ACTION,
    payload: selectedIndex
  };
}

export function scrollSelectorMouseAction(mouseX: number): MouseMoveAct {
  return {
    type: MOUSE_ACTION,
    payload: mouseX
  };
}

export type ScrollAction = ScrollSelectorAct | MouseMoveAct;
