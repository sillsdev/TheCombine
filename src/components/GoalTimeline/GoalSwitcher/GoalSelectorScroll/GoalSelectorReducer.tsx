import {
  GoalScrollAction,
  SELECT_ACTION,
  MOUSE_ACTION
} from "./GoalSelectorAction";
import { GoalSelectorState } from "../../../../types/goals";

export const defaultState: GoalSelectorState = {
  selectedIndex: 0,
  goalOptions: [],
  mouseX: 0,
  lastIndex: 0
};

export const goalSelectReducer = (
  state: GoalSelectorState | undefined,
  action: GoalScrollAction
): GoalSelectorState => {
  if (!state) return defaultState;
  switch (action.type) {
    case SELECT_ACTION:
      return {
        ...state,
        selectedIndex: action.payload
      };

    case MOUSE_ACTION:
      return {
        ...state,
        mouseX: action.payload
      };

    default:
      return state;
  }
};
