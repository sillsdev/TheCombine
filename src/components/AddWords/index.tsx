import AddWords from "./AddWordsComponent";
import { connect } from "react-redux";
import { StoreState } from "../../types";
import { getTranslate } from "react-localize-redux";

function mapStateToProps(state: StoreState) {
  return {
    translate: getTranslate(state.localize)
  };
}

export default connect(
  mapStateToProps,
  null
)(AddWords);
