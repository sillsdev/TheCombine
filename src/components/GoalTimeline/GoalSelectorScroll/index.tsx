import { connect } from "react-redux";

import { StoreState } from "../../../types";
import { StoreStateDispatch } from "../../../types/actions";
import { GoalSelectorState } from "../../../types/goals";
import {
  scrollSelectorIndexAction,
  scrollSelectorMouseAction,
} from "./GoalSelectorAction";
import GoalSelectorScroll from "./GoalSelectorScroll";

export function mapStateToProps(state: StoreState): GoalSelectorState {
  return {
    allPossibleGoals: state.goalsState.allPossibleGoals,
    selectedIndex: state.goalSelectorState.selectedIndex,
    mouseX: state.goalSelectorState.mouseX,
    lastIndex: state.goalsState.allPossibleGoals.length - 1,
  };
}

export function mapDispatchToProps(dispatch: StoreStateDispatch) {
  return {
    swapSelectedIndex: (ndx: number) => {
      dispatch(scrollSelectorIndexAction(ndx));
    },
    swapMouseX: (iX: number) => {
      dispatch(scrollSelectorMouseAction(iX));
    },
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(GoalSelectorScroll);
