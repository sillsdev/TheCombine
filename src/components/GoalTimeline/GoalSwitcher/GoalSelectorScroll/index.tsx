import GoalSelectorScroll from "./GoalSelectorScroll";
import {
  GoalScrollAction,
  scrollSelectorIndexAction,
  scrollSelectorMouseAction,
} from "./GoalSelectorAction";

import { connect } from "react-redux";
import { StoreState } from "../../../../types";
import { GoalSelectorState } from "../../../../types/goals";
import { Dispatch } from "redux";

export function mapStateToProps(state: StoreState): GoalSelectorState {
  return {
    allPossibleGoals: state.goalsState.allPossibleGoals,
    selectedIndex: state.goalSelectorState.selectedIndex,
    mouseX: state.goalSelectorState.mouseX,
    lastIndex: state.goalsState.allPossibleGoals.length - 1,
  };
}

export function mapDispatchToProps(dispatch: Dispatch<GoalScrollAction>) {
  return {
    swapSelectedIndex: (ndx: number) => {
      dispatch(scrollSelectorIndexAction(ndx));
    },
    swapMouseX: (iX: number) => {
      dispatch(scrollSelectorMouseAction(iX));
    },
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(GoalSelectorScroll);
