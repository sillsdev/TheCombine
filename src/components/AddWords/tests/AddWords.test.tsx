import React from "react";

import GoalSelectorScroll from "..";

import { Provider } from "react-redux";
import { act, Simulate } from "react-dom/test-utils";
import renderer, {
  ReactTestInstance,
  ReactTestRenderer
} from "react-test-renderer";
import AddWords from "../";
import configureStore from "redux-mock-store";
import AddWords_unconnected from "../AddWordsComponent";
import { LocalizeProvider } from "react-localize-redux";
import configureMockStore from "redux-mock-store";
import createMockStore from "redux-mock-store";
import { GoalSelectorState } from "../../../types/goals";
import axios from "axios";
import { Word } from "../../../types/word";

let master: renderer.ReactTestRenderer;
let handle: renderer.ReactTestInstance;

// Mock getTranslate
const MOCK_TRANSLATE = jest.fn(_ => {
  return "dummy";
});
jest.mock("react-localize-redux", () => {
  const localize = jest.requireActual("react-localize-redux");
  return {
    ...localize,
    getTranslate: jest.fn(_ => {
      return MOCK_TRANSLATE;
    })
  };
});
jest.mock("axios");
let mockedAxios = axios as jest.Mocked<typeof axios>;

// Mock store
const mockStore = configureStore();

beforeEach(() => {
  // Here, use the act block to be able to render our AddWords into the DOM
  // Re-created each time to prevent actions from previous runs from affecting future runs
  act(() => {
    master = renderer.create(
      <AddWords_unconnected domain={"en"} translate={jest.fn(() => "ok")} />
    );
    handle = master.root.findByType(AddWords_unconnected);
  });

  mockedAxios.put.mockClear();
});

describe("Tests AddWords", () => {
  it("Constructs correctly", () => {
    // Default snapshot test
    snapTest("default view");
  });

  it("Adds a word", done => {
    handle.instance.setState({ newVern: "testVern", newGloss: "testGloss" });
    mockedAxios.post.mockImplementationOnce((url, word: Word) => {
      return Promise.resolve({ data: "123" });
    });
    handle.instance.submit(undefined, () => {
      expect(handle.instance.state.rows).toEqual([
        {
          vernacular: "testVern",
          glosses: "testGloss",
          id: "123"
        }
      ]);
      done();
    });
  });

  it("Edits a word", done => {
    handle.instance.setState({
      rows: [
        {
          vernacular: "testVern1",
          glosses: "testGloss1",
          id: "123"
        },
        {
          vernacular: "testVern2",
          glosses: "testGloss2",
          id: "456"
        },
        {
          vernacular: "testVern3",
          glosses: "testGloss3",
          id: "789"
        }
      ]
    });
    mockedAxios.put.mockResolvedValue(1);
    handle.instance.updateWord(1);
    expect(mockedAxios.put).toHaveBeenCalledTimes(1);
    done();
  });

  it("Removes a word", done => {
    handle.instance.setState({
      rows: [
        {
          vernacular: "testVern1",
          glosses: "testGloss1",
          id: "123"
        },
        {
          vernacular: "testVern2",
          glosses: "testGloss2",
          id: "456"
        },
        {
          vernacular: "testVern3",
          glosses: "testGloss3",
          id: "789"
        }
      ]
    });
    mockedAxios.delete.mockResolvedValue(1);
    handle.instance.removeWord(1, () => {
      expect(handle.instance.state.rows).toEqual([
        {
          vernacular: "testVern1",
          glosses: "testGloss1",
          id: "123"
        },
        {
          vernacular: "testVern3",
          glosses: "testGloss3",
          id: "789"
        }
      ]);
      done();
    });
  });
});

// Utility functions -----------------------------

// Perform a snapshot test
function snapTest(name: string) {
  expect(master.toJSON()).toMatchSnapshot();
}
