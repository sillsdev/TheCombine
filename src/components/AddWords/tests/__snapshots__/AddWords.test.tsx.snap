// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests AddWords Constructs correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "marginLeft": "auto",
      "marginRight": "auto",
      "maxWidth": 800,
      "padding": 32,
    }
  }
>
  <h4
    className="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter"
    style={
      Object {
        "marginBottom": 32,
      }
    }
  >
    Missing translationId: addWords.domain for language: \${ languageCode }
    : 
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <h4
          className="MuiTypography-root MuiTypography-h4"
        >
          en (1)
        </h4>
      </span>
    </button>
  </h4>
  <hr
    className="MuiDivider-root"
  />
  <form
    onSubmit={[Function]}
  >
    <input
      style={
        Object {
          "display": "none",
        }
      }
      type="submit"
    />
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
      >
        <h5
          className="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
          style={
            Object {
              "marginTop": 32,
            }
          }
        >
          Missing translationId: addWords.vernacular for language: \${ languageCode }
        </h5>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
      >
        <h5
          className="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
          style={
            Object {
              "marginTop": 32,
            }
          }
        >
          Missing translationId: addWords.glosses for language: \${ languageCode }
        </h5>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div
          className="MuiGrid-root MuiGrid-container"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
            style={
              Object {
                "paddingLeft": 32,
                "paddingRight": 32,
              }
            }
          >
            <div
              className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              onKeyDown={[Function]}
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink={false}
              >
                Missing translationId: addWords.vernacular for language: \${ languageCode }
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                onClick={[Function]}
              >
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-246 MuiOutlinedInput-notchedOutline"
                  style={
                    Object {
                      "paddingLeft": 8,
                    }
                  }
                >
                  <legend
                    className="PrivateNotchedOutline-legend-247"
                    style={
                      Object {
                        "width": 0.01,
                      }
                    }
                  >
                    <span
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&#8203;",
                        }
                      }
                    />
                  </legend>
                </fieldset>
                <input
                  aria-invalid={false}
                  autoFocus={true}
                  className="MuiInputBase-input MuiOutlinedInput-input"
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
            style={
              Object {
                "paddingLeft": 32,
                "paddingRight": 32,
              }
            }
          >
            <div
              className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              onKeyDown={[Function]}
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink={false}
              >
                Missing translationId: addWords.glosses for language: \${ languageCode }
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                onClick={[Function]}
              >
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-246 MuiOutlinedInput-notchedOutline"
                  style={
                    Object {
                      "paddingLeft": 8,
                    }
                  }
                >
                  <legend
                    className="PrivateNotchedOutline-legend-247"
                    style={
                      Object {
                        "width": 0.01,
                      }
                    }
                  >
                    <span
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&#8203;",
                        }
                      }
                    />
                  </legend>
                </fieldset>
                <input
                  aria-invalid={false}
                  className="MuiInputBase-input MuiOutlinedInput-input"
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-flex-end"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "marginTop": 32,
          }
        }
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          Missing translationId: addWords.next for language: \${ languageCode }
        </span>
      </button>
    </div>
  </div>
</div>
`;
